File: python/day11.py
def valid_pass(password):
    alphabet = "abcdefghijklmnopqrstuvwxyz"
    triples = ["".join(aas) for aas in zip(alphabet, alphabet[1:], alphabet[2:])]
    pairs = ["".join(aas) for aas in zip(alphabet, alphabet)]

    forbidden = ["i", "o", "l"]

    for f in forbidden:
        if f in password:
            return False

    if not any([t in password for t in triples]):
        return False

    count = 0
    for p in pairs:
        if p in password:
            count += 1

    if count < 2:
        return False

    return True


def increment_pass(password):
    ps = [ord(p) - ord('a') for p in password]

    for i in range(len(ps) - 1, -1, -1):
        ps[i] += 1
        if ps[i] <= 25:
            break
        ps[i] -= 26

    # Skip past any forbidden letters
    forbidden = [ord(c) - ord('a') for c in "iol"]

    for i, p in enumerate(ps):
        if p in forbidden:
            ps[i] += 1
            for j in range(i+1, len(ps)):
                ps[j] = 0

    return "".join([chr(p + ord('a')) for p in ps])

def next_pass(password):
    while not valid_pass(password):
        password = increment_pass(password)
    return password

n = next_pass("cqjxjnds")
print(n)
print(next_pass(increment_pass(n)))

File: aoc2015/src/bin/day11.rs
const ALPHABET: &str = "abcdefghijklmnopqrstuvwxyz";
const INITIAL_PASSWORD: &str = "cqjxjnds";
const FORBIDDEN: &str = "iol";

struct Validator {
    triples: Vec<String>,
    pairs: Vec<String>,
    forbidden: Vec<String>,
}

impl Validator {
    fn new() -> Self {
        Validator{
            triples: ALPHABET.chars().zip(ALPHABET.chars().skip(1)).zip(ALPHABET.chars().skip(2)).map(|((x,y),z)| x.to_string()+&y.to_string()+&z.to_string()).collect(),
            pairs: ALPHABET.chars().zip(ALPHABET.chars()).map(|(x,y)| x.to_string()+&y.to_string()).collect(),
            forbidden: FORBIDDEN.chars().map(|x| x.to_string()).collect(),
        }
    }

    fn validate(&self, p: &str) -> bool {
        if self.forbidden.iter().map(|f| p.contains(f)).any(|x| x) {
            return false;
        }

        if !self.triples.iter().map(|t| p.contains(t)).any(|x| x) {
            return false;
        }

        let count: u32 = self.pairs.iter().map(|x| if p.contains(x) { 1 } else { 0 }).sum();

        if count < 2 {
            return false;
        }

        true
    }
}

fn increment_pass(p: &str) -> String {
    let mut ps: Vec<_> = p.chars().map(|x| x as u32 - 'a' as u32).collect();

    for i in (0..ps.len()).rev() {
        ps[i] += 1;
        if ps[i] <= 25 {
            break;
        }
        ps[i] -= 26;
    }

    ps.into_iter().map(|p| std::char::from_u32(p + 'a' as u32).unwrap()).collect()
}

fn main() {
    let v = Validator::new();

    let mut p = INITIAL_PASSWORD.to_owned();
    while !v.validate(&p) {
        p = increment_pass(&p);
    }
    println!("Part 1: {}", p);
    p = increment_pass(&p);
    while !v.validate(&p) {
        p = increment_pass(&p);
    }
    println!("Part 2: {}", p);
}
